product_category_code_unique/models/product_category.py

```diff
@@ -1,26 +1,73 @@
 # Copyright 2021 ACSONE SA/NV
+# Copyright 2024 Manuel Regidor <manuel.regidor@sygel.es>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import api, models
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
 
 
 class ProductCategory(models.Model):
 
     _inherit = "product.category"
 
-    _sql_constraints = [
-        ("uniq_code", "unique(code)", "The category code must be unique!"),
-    ]
+    def _get_parents(self):
+        self.ensure_one()
+        return self.search([("parent_id", "parent_of", self.id)])
+
+    def _get_children(self):
+        self.ensure_one()
+        return self.search([("parent_id", "child_of", self.id)])
+
+    def _get_hierarchy_cats(self):
+        self.ensure_one()
+        return self._get_parents() + self._get_children()
+
+    def _code_restriction(self, restriction=False):
+        restriction = restriction or self.env["ir.config_parameter"].get_param(
+            "product_code_unique.product_code_unique_restriction"
+        )
+        if restriction:
+            for cat in self:
+                domain = []
+                within = ""
+                if restriction == "system":
+                    domain = [("code", "=", cat.code)]
+                    within = _("the system")
+                elif restriction == "direct":
+                    to_check = cat.parent_id + cat.child_id
+                    domain = [
+                        "|",
+                        ("id", "=", cat.id),
+                        "&",
+                        ("code", "=", cat.code),
+                        ("id", "in", to_check.ids),
+                    ]
+                    within = _("parent and children")
+                elif restriction == "hierarchy":
+                    to_check = cat._get_hierarchy_cats()
+                    domain = [("code", "=", cat.code), ("id", "in", to_check.ids)]
+                    within = _("category hierarchy")
+                if self.search_count(domain) > 1:
```

Comment:
IMHO, this should be avoided for performance reasons. 

A query to get all the categories should be done outside the loop. Then, compare the recordset values.

Reply:
Sorry, but I don't know exactly what you mean here.

Reply:
@rousseldenis I've checked your commet again but I still don't know exactly what changes you are suggesting here. Please, could you give me a more detailed explanation so I can do the changes? Thank you

Reply:
self = recordset. So, the loop you do will launch an SQL query for each `search_count` instance.

For performances reasons, this should be avoided. You should find a way to get all results before the loop. Then do the logic with the result.

Reply:
I've applied some changes. Could you have a look? Thank you.